components/script/dom/globalscope.rs

```diff
@@ -2848,6 +2849,28 @@ impl GlobalScope {
         is_js_evaluation_allowed == CheckResult::Allowed
     }
 
+    pub(crate) fn should_navigation_request_be_blocked(&self, url: &ServoUrl) -> bool {
+        let Some(csp_list) = self.get_csp_list() else {
+            return false;
+        };
+        let request = Request {
+            url: url.clone().into_url(),
+            origin: url.origin().into_url_origin(),
```

Comment:
We need to use the origin of the global that initiated the navigation, not the origin of the URL: https://www.w3.org/TR/CSP/#match-request-to-source-list . This is LoadData::load_origin.